File: internal/storage/metrics.go
package storage

import (
	"time"
)

// SystemMetric 系统指标历史记录
type SystemMetric struct {
	ID        uint      `gorm:"primaryKey" json:"id"`
	AgentID   string    `gorm:"index;not null" json:"agentId"`
	Timestamp time.Time `gorm:"index" json:"timestamp"`

	// CPU 信息
	CPUUsage      float64 `json:"cpuUsage"`       // CPU 使用率 (0-100)
	CPUCores      int     `json:"cpuCores"`       // CPU 核心数
	CPUFreqMHz    float64 `json:"cpuFreqMhz"`     // CPU 频率

	// 内存信息
	MemTotal      uint64  `json:"memTotal"`       // 总内存 (bytes)
	MemUsed       uint64  `json:"memUsed"`        // 已用内存 (bytes)
	MemAvailable  uint64  `json:"memAvailable"`   // 可用内存 (bytes)
	MemUsagePercent float64 `json:"memUsagePercent"` // 内存使用率

	// 磁盘信息
	DiskTotal     uint64  `json:"diskTotal"`      // 总磁盘空间 (bytes)
	DiskUsed      uint64  `json:"diskUsed"`       // 已用磁盘空间 (bytes)
	DiskFree      uint64  `json:"diskFree"`       // 可用磁盘空间 (bytes)
	DiskUsagePercent float64 `json:"diskUsagePercent"` // 磁盘使用率

	// 网络信息
	NetBytesSent  uint64  `json:"netBytesSent"`   // 发送字节数
	NetBytesRecv  uint64  `json:"netBytesRecv"`   // 接收字节数

	// 系统信息
	OSName        string  `json:"osName"`         // 操作系统名称
	OSVersion     string  `json:"osVersion"`      // 操作系统版本
	Arch          string  `json:"arch"`           // 架构 (amd64, arm64等)
	Uptime        uint64  `json:"uptime"`         // 系统运行时间 (seconds)

	// 负载信息 (Unix-like)
	Load1         float64 `json:"load1"`          // 1分钟负载
	Load5         float64 `json:"load5"`          // 5分钟负载
	Load15        float64 `json:"load15"`         // 15分钟负载

	CreatedAt     time.Time `json:"createdAt"`
}

// SystemInfoSnapshot 系统信息快照（最新状态）
type SystemInfoSnapshot struct {
	ID            uint      `gorm:"primaryKey" json:"id"`
	AgentID       string    `gorm:"uniqueIndex;not null" json:"agentId"`

	// CPU 信息
	CPUUsage      float64   `json:"cpuUsage"`
	CPUCores      int       `json:"cpuCores"`
	CPUFreqMHz    float64   `json:"cpuFreqMhz"`

	// 内存信息
	MemTotal      uint64    `json:"memTotal"`
	MemUsed       uint64    `json:"memUsed"`
	MemAvailable  uint64    `json:"memAvailable"`
	MemUsagePercent float64 `json:"memUsagePercent"`

	// 磁盘信息
	DiskTotal     uint64    `json:"diskTotal"`
	DiskUsed      uint64    `json:"diskUsed"`
	DiskFree      uint64    `json:"diskFree"`
	DiskUsagePercent float64 `json:"diskUsagePercent"`

	// 网络信息
	NetBytesSent  uint64    `json:"netBytesSent"`
	NetBytesRecv  uint64    `json:"netBytesRecv"`

	// 系统信息
	OSName        string    `json:"osName"`
	OSVersion     string    `json:"osVersion"`
	Arch          string    `json:"arch"`
	Uptime        uint64    `json:"uptime"`
	Hostname      string    `json:"hostname"`

	// 负载信息
	Load1         float64   `json:"load1"`
	Load5         float64   `json:"load5"`
	Load15        float64   `json:"load15"`

	UpdatedAt     time.Time `json:"updatedAt"`
}

// TableName 指定表名
func (SystemMetric) TableName() string {
	return "system_metrics"
}

// TableName 指定表名
func (SystemInfoSnapshot) TableName() string {
	return "system_info_snapshots"
}

File: web/src/services/metrics.ts
import axios from 'axios';
import { logger } from '@/utils/logger';

// 系统信息快照
export interface SystemInfoSnapshot {
  id: number;
  agentId: string;
  cpuUsage: number;
  cpuCores: number;
  cpuFreqMhz: number;
  memTotal: number;
  memUsed: number;
  memAvailable: number;
  memUsagePercent: number;
  diskTotal: number;
  diskUsed: number;
  diskFree: number;
  diskUsagePercent: number;
  netBytesSent: number;
  netBytesRecv: number;
  osName: string;
  osVersion: string;
  arch: string;
  uptime: number;
  hostname: string;
  load1: number;
  load5: number;
  load15: number;
  updatedAt: string;
}

// 系统指标历史记录
export interface SystemMetric {
  id: number;
  agentId: string;
  timestamp: string;
  cpuUsage: number;
  cpuCores: number;
  cpuFreqMhz: number;
  memTotal: number;
  memUsed: number;
  memAvailable: number;
  memUsagePercent: number;
  diskTotal: number;
  diskUsed: number;
  diskFree: number;
  diskUsagePercent: number;
  netBytesSent: number;
  netBytesRecv: number;
  osName: string;
  osVersion: string;
  arch: string;
  uptime: number;
  load1: number;
  load5: number;
  load15: number;
  createdAt: string;
}

// 历史指标响应
export interface MetricsHistoryResponse {
  data: SystemMetric[];
  start: number;
  end: number;
  count: number;
}

// 创建 axios 实例
const metricsClient = axios.create({
  baseURL: '/api',
  timeout: 30000,
  headers: {
    'Content-Type': 'application/json',
  },
});

// 请求拦截器 - 添加 token
metricsClient.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('token');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

// 响应拦截器
metricsClient.interceptors.response.use(
  (response) => response.data,
  (error) => {
    if (error.response?.status === 401) {
      localStorage.removeItem('token');
      localStorage.removeItem('username');
      window.location.href = '/login';
    }
    logger.error('Metrics API Error:', error);
    return Promise.reject(error);
  }
);

// 获取所有系统信息快照
export async function getSystemSnapshots(): Promise<SystemInfoSnapshot[]> {
  const response = await metricsClient.get<unknown, { data: SystemInfoSnapshot[] }>('/metrics/snapshots');
  return response?.data || [];
}

// 获取单个 Agent 的系统信息
export async function getSystemSnapshot(agentId: string): Promise<SystemInfoSnapshot> {
  return metricsClient.get<unknown, SystemInfoSnapshot>(`/metrics/snapshot?agent_id=${agentId}`);
}

// 获取历史指标
export async function getMetricsHistory(
  agentId: string,
  params?: {
    start?: number;
    end?: number;
    limit?: number;
  },
): Promise<MetricsHistoryResponse> {
  const response = await metricsClient.get<unknown, MetricsHistoryResponse>(
    `/metrics/history?agent_id=${agentId}`,
    { params }
  );
  return {
    data: response?.data || [],
    start: response?.start || 0,
    end: response?.end || 0,
    count: response?.count || 0,
  };
}

// 格式化字节数

// 格式化百分比

// 格式化运行时间
